File: srclab-common/src/main/java/xyz/srclab/common/reflect/TypeRef.java
package xyz.srclab.common.reflect;

import xyz.srclab.annotation.Immutable;
import xyz.srclab.annotation.Nullable;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

@Immutable
public class TypeRef<T> {

    public static <T> TypeRef<T> of(Type type) {
        return new TypeRef<>(type);
    }

    private final Type type;

    protected TypeRef() {
        this.type = getGenericSuperclass();
    }

    private TypeRef(Type type) {
        this.type = type;
    }

    protected Type getGenericSuperclass() {
        @Nullable Type generic = TypeKit.getGenericSuperclass(getClass(), TypeRef.class);
        if (!(generic instanceof ParameterizedType)) {
            throw new IllegalStateException("Generic super class must be a parameterized type");
        }
        ParameterizedType parameterizedType = (ParameterizedType) generic;
        return parameterizedType.getActualTypeArguments()[0];
    }

    public Type getType() {
        return type;
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/reflect/TypeRef.kt
package xyz.srclab.common.reflect

import xyz.srclab.common.aboutBoat
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

abstract class TypeRef<T> {

    @get:JvmName("type")
    val type: Type

    protected constructor() {
        type = reflectGenericSuperclass()
    }

    protected constructor(type: Type) {
        this.type = type
    }

    private fun reflectGenericSuperclass(): Type {
        val generic = this::class.java.genericSuperClassFor(TypeRef::class.java)
        if (generic !is ParameterizedType) {
            throw IllegalStateException(
                "Reflect to generic superclass failed, it seems a bug, please report to ${aboutBoat().report}"
            )
        }
        return generic.actualTypeArguments[0]
    }

    override fun toString(): String {
        return "TypeRef($type)"
    }

    companion object {

        @JvmStatic
        fun <T> of(type: Type): TypeRef<T> {
            return object : TypeRef<T>(type) {}
        }
    }
}

val <T> T.typeRef: TypeRef<T>
    get() {
        return object : TypeRef<T>() {}
    }